core/system/src/context.c

```diff
@@ -0,0 +1,266 @@
+/*
+ * Copyright (c) 2013-2016, ARM Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <uvisor.h>
+#include "context.h"
+#include "secure_gateway.h"
+#include "svc.h"
+#include "vmpu.h"
+#include "debug.h"
+
+/* Currently active box */
+uint8_t g_active_box;
+
+/** Stack of previous context states
+ *
+ * @internal
+ *
+ * Each element in this array holds information about a previously active
+ * context. The stack grows downwards when context switches are nested, and
+ * upwards when nested context switches return. */
+TContextPreviousState g_context_previous_states[UVISOR_CONTEXT_MAX_DEPTH];
+
+/** Stack pointer for the stack of previous contexts
+ *
+ * @internal */
+int g_context_p;
+
+/** Pushes the previous state to the state stack and updates the current state.
+ *
+ * @internal
+ *
+ * @warning This function trusts all the arguments that are passed to it. Input
+ * verification should be performed by the caller.
+ *
+ * @param context_type[in]  Type of context switch to perform
+ * @param src_id[in]        ID of the box we are switching context from
+ * @param src_sp[in]        Stack pointer of the box we are switching from */
+static void context_state_push(TContextSwitchType context_type, uint8_t src_id, uint32_t src_sp)
+{
+    /* Check that the state stack does not overflow. */
+    if (g_context_p == UVISOR_CONTEXT_MAX_DEPTH) {
```

Comment:
What about underflow? `g_context_p` is currently an `int`.

What about overflow beyond `UVISOR_CONTEXT_MAX_DEPTH`?